File: ruby/sinatra/app/controllers/tasks_controller.rb
class TasksController < ApplicationController
  def index
    response.status = 200
    current_user.tasks.to_json
  end

  def create
    @task = current_user.tasks.new(task_params)

    if @task.save
      response.status = 201
      @task.to_json
    else
      response.status = 422
      { errors: task_errors }.to_json
    end
  end

  def destroy
    set_task

    if @task.present?
      @task.destroy
      response.status = 200
    else
      response.status = 404
      { errors: { task: "Task must exist" } }.to_json
    end
  end

  def complete
    set_task

    if @task.present?
      @task.update(completed: true)
      response.status = 200
      @task.to_json
    else
      response.status = 404
      { errors: { task: "Task must exist" } }.to_json
    end
  end

  private

  def set_task
    @task = current_user.tasks.find_by(id: params[:id])
  end

  def task_errors
    @task.errors.map { |error| [error.attribute, error.full_message] }.to_h
  end

  def task_params
    params[:task]&.slice(:title) || {}
  end
end

File: node/express/src/controllers/tasks.controller.ts
import type { RequestHandler, Response } from "express"
import { z as zod } from "zod"
import { Prisma } from "@prisma/client"
import prisma from "src/models/task.model"

export const index: RequestHandler = async (_, res) => {
  const tasks = await prisma.task.findMany()
  return res.status(200).json(tasks)
}

export const create: RequestHandler = async (req, res) => {
  const { task: data } = req.body as { task: Prisma.TaskUncheckedCreateInput }

  try {
    const task = await prisma.task.create({ data: { ...data, userId: req.currentUser.id } })
    return res.status(201).json(task)
  } catch (error) {
    return _handleError(error, res)
  }
}

export const destroy: RequestHandler = async (req, res) => {
  try {
    await prisma.task.delete({ where: { id: +req.params.id } })
    return res.status(200).send()
  } catch (error) {
    return _handleError(error, res)
  }
}

export const complete: RequestHandler = async (req, res) => {
  try {
    const task = await prisma.task.update({ data: { completed: true }, where: { id: +req.params.id } })
    return res.status(200).json(task)
  } catch (error) {
    return _handleError(error, res)
  }
}

const _handleError = (error: unknown, res: Response) => {
  if (error instanceof zod.ZodError) {
    return res.status(422).json({
      errors: Object.fromEntries(error.issues.map(issue => [issue.path[0], issue.message]))
    })
  }

  if (error instanceof Prisma.PrismaClientKnownRequestError) {
    if (error.code === "P2025") {
      // "An operation failed because it depends on one or more records that were required but not found."
      return res.status(404).json({ errors: { task: "Task must exist" } })
    }
  }

  throw error
}
